File: Python/project-euler/problems/problem_018.py
from utils.code_runner import execute_code


def problem():
    with open("../resources/problem_018.txt") as file:
        digits = [line.rstrip('\n') for line in file]

    for i in range(len(digits)):
        digits[i] = [int(x) for x in digits[i].strip().split(' ')]
    # Iterate from the digits array by -1 until 0
    for x in range(len(digits) - 1, 0, -1):
        for y in range(0, x):
            # Get the highest sum of the two possible paths
            digits[x - 1][y] += max(digits[x][y], digits[x][y + 1])

    total_sum = digits[0][0]

    return total_sum


if __name__ == '__main__':
    execute_code(problem)

File: Java/src/main/java/nl/davydehaas/projecteuler/Problem018.java
package nl.davydehaas.projecteuler;

import static nl.davydehaas.projecteuler.util.SolutionTimer.timeSolution;

import java.util.Arrays;

import nl.davydehaas.projecteuler.util.InputReader;

public class Problem018 {

    void main() {
        timeSolution(Problem018::solve);
    }

    static int solve() {
        int[][] pyramid = InputReader.readFile(18).stream()
                .map(line -> Arrays.stream(line.split(" "))
                        .mapToInt(Integer::parseInt)
                        .toArray())
                .toArray(int[][]::new);

        // Iterate from the digits array by -1 until 0
        for (int x = pyramid.length - 1; x > 0; x--) {
            for (int y = 0; y < x; y++) {
                // Get the highest sum of the two possible paths.
                pyramid[x - 1][y] += Math.max(pyramid[x][y], pyramid[x][y + 1]);
            }
        }

        return pyramid[0][0];
    }
}
